packages/medusa/src/services/batch-job.ts

```diff
@@ -71,6 +85,174 @@ class BatchJobService extends TransactionBaseService<BatchJobService> {
+    return await this.atomicPhase_(async (manager) => {
+      let batchJob: BatchJob = batchJobOrId as BatchJob
+      if (typeof batchJobOrId === "string") {
+        batchJob = await this.retrieve(batchJobOrId)
+      }
+
+      await this.eventBus_
+        .withTransaction(manager)
+        .emit(BatchJobService.Events.PROCESS_COMPLETE, {
+          id: batchJob.id,
+        })
+
+      return batchJob
+    })
+  }
+
+  async validateBatchContext(
+    type: string,
+    context: Record<string, unknown>
+  ): Promise<Record<string, unknown>> {
+    // TODO validate context with batch job strategy corresponding to the given type
+
+    return context
+  }
+
+  async create(data: BatchJobCreateProps): Promise<BatchJob> {
+    return await this.atomicPhase_(async (manager) => {
+      const batchJobRepo: BatchJobRepository = manager.getCustomRepository(
+        this.batchJobRepository_
+      )
+
+      // TODO use strategy to validate the context
+
+      const validatedContext = await this.validateBatchContext(
+        data.type,
+        data.context
+      )
+      const toCreate = {
+        status: BatchJobStatus.CREATED,
+        ...data,
+        context: validatedContext,
+      } as DeepPartial<BatchJob>
+
+      const batchJob = await batchJobRepo.create(toCreate)
+      const result = await batchJobRepo.save(batchJob)
+
+      await this.eventBus_
+        .withTransaction(manager)
+        .emit(BatchJobService.Events.CREATED, {
+          id: result.id,
+        })
+
+      return result
+    })
+  }
+
+  async update(
+    batchJobId: string,
+    data: BatchJobUpdateProps
+  ): Promise<BatchJob> {
+    return await this.atomicPhase_(async (manager) => {
+      const batchJobRepo: BatchJobRepository = manager.getCustomRepository(
+        this.batchJobRepository_
+      )
+
+      const batchJob = await this.retrieve(batchJobId)
+
+      const { status, ...rest } = data
+
+      if (typeof status !== "undefined") {
+        batchJob.status = status
+
+        switch (status) {
+          case BatchJobStatus.PROCESSING:
+            batchJob.processing_at = new Date()
+            break
+          case BatchJobStatus.AWAITING_CONFIRMATION:
+            batchJob.awaiting_confirmation_at = new Date()
+            break
+          case BatchJobStatus.COMPLETED:
+            batchJob.completed_at = new Date()
+            break
+          case BatchJobStatus.CANCELED:
+            batchJob.canceled_at = new Date()
+            break
+        }
+      }
+
+      Object.keys(rest)
+        .filter((restKey) => typeof rest[restKey] !== `undefined`)
+        .forEach((restKey) => {
+          batchJob[restKey] = rest[restKey]
+        })
+
+      return await batchJobRepo.save(batchJob)
+    })
+  }
+
+  /*
+   * if job is started with dry_run: true, then it's required
+   * to complete the job before it's written to DB
+   */
+  async complete(batchJobOrId: string | BatchJob): Promise<BatchJob | never> {
```

Comment:
We could consider if we should keep this or not if we remove the endpoint. Workers could potentially use this method for completing the batch jobs and it would provide a standardized way of completing jobs to ensure a bit of integrity (mainly status)

Reply:
I think that indeed we will keep that method to be used by the worker

Reply:
At the moment we cant update the status to `ready` and a couple of other batch statuses. We should probably create a method like this for each of the missing status updates: `ready`, `processing`, `awaiting_confirmation` and `failed` in order to update the correct timestamp and emit the corresponding event (which for most of these should be added as well). Alternatively we should include certain timestamp updates in the `update` method and emit events based on those when relevant.

wdyt?

Reply:
yeah, it is a bit confusing,
the flow is the following
```
dry_run -> processing -> summary -> awaiting_confirmation -> confirm -> write -> complete
!dry_run -> processing -> write -> complete
```
If I am not mistaken

Reply:
i ll remove the comment that is old now